src/lua/utils.c

```diff
@@ -328,7 +342,43 @@ luaL_register_module(struct lua_State *L, const char *modname,
 			luaL_error(L, "Failed to register library");
 	}
 	lua_remove(L, -2);  /* remove _LOADED table */
-	luaL_register(L, NULL, methods);
+	luaT_setfuncs(L, methods);
+}
+
+int
+luaT_newmodule(struct lua_State *L, const char *modname,
+	       const struct luaL_Reg *funcs)
+{
+	say_debug("%s(%s)", __func__, modname);
+	assert(modname != NULL && funcs != NULL);
+
+	/* Get package.loaded. */
+	lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
+
+	/* Verify that the module is not already registered. */
+	lua_getfield(L, -1, modname);
+	assert(lua_isnil(L, -1));
+	lua_pop(L, 1);
+
+	/* Create a module table. */
+	lua_newtable(L);
+
+	/* Fill the module table with functions. */
+	luaT_setfuncs(L, funcs);
+
+	/* Copy the module table. */
+	lua_pushvalue(L, -1);
+
+	/* package.loaded[modname] = <module table> */
+	lua_setfield(L, -3, modname);
+
+	/* Stack: package.loaded, module table. */
+
+	/* Drop package.loaded. Stack: module table. */
+	lua_remove(L, -2);
```

Comment:
Lua will do this automatically, later (see `return 1` below): there is no need in this line.

Reply:
Discussed with @Totktonada: this `lua_remove` is required here, since `luaT_newmodule` is used mostly in C, hence the guest stack should be consistent after its return.

Ask Sasha to drop a comment regarding this.

Reply:
I looked on it once again. The function can't be called from Lua directly, because it has arguments beyond `struct lua_State *`. So it would be a bit strange to highlight a fact, which is anyway considered as a [good practice](https://www.lua.org/manual/5.1/manual.html#lua_call). And it is part of the function contract.

Anyway, I changed the comment to make it more obvious that I'm doing that deliberately:

```c
/*
 * Drop package.loaded to leave the stack in its original
 * state.
 */
lua_remove(L, -2);
```

Reply:
Thanks a huge!